File: support-frontend/src/types/ticket.ts
/**
 * Ticket-related TypeScript types.
 */

export type TicketStatus = "pending" | "in_progress" | "paused" | "in_review" | "completed";
export type TicketType = "technical_error" | "bug" | "feature" | "remove";
export type Priority = "critical" | "high" | "medium" | "low";

export interface Ticket {
  id: string;
  project_id: string;
  title: string;
  description: string;
  type: TicketType;
  priority: Priority;
  status: TicketStatus;
  jira_key: string | null;
  /** Bare FK column: the creator's user id. */
  created_by: string;
  /** Bare FK column: the assignee's user id (null if unassigned). */
  assigned_to: string | null;
  /** Nested creator user object (embedded by the API via a join). */
  created_by_user: UserSummary | null;
  /** Nested assignee user object (embedded by the API via a join). */
  assigned_to_user: UserSummary | null;
  attachments_count: number;
  activities_count: number;
  created_at: string;
  updated_at: string;
}

export interface TicketCreate {
  title: string;
  description: string;
  type: TicketType;
  priority: Priority;
  assigned_to?: string;
}

export interface TicketUpdate {
  title?: string;
  description?: string;
  type?: TicketType;
  priority?: Priority;
  status?: TicketStatus;
  assigned_to?: string | null;
}

export interface UserSummary {
  id: string;
  name: string;
  email: string;
}

export interface Attachment {
  id: string;
  ticket_id: string;
  file_name: string;
  s3_key: string;
  content_type: string;
  file_size: number;
  download_url: string | null;
  uploaded_by_user?: UserSummary | null;
  uploaded_at: string;
}

export type ActivityActionType =
  | "created"
  | "status_changed"
  | "updated"
  | "commented"
  | "file_uploaded"
  | "file_deleted"
  | "assigned";

export interface Activity {
  id: string;
  ticket_id: string;
  action_type: ActivityActionType;
  actor: UserSummary | null;
  old_value: Record<string, unknown> | null;
  new_value: Record<string, unknown> | null;
  comment: string | null;
  created_at: string;
}

File: support-backend/app/schemas/ticket.py
"""Ticket request/response schemas with full validation."""

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


class TicketStatus(str, Enum):
    """Ticket status enum."""

    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    PAUSED = "paused"
    IN_REVIEW = "in_review"
    COMPLETED = "completed"


class TicketType(str, Enum):
    """Ticket type enum."""

    TECHNICAL_ERROR = "technical_error"
    BUG = "bug"
    FEATURE = "feature"
    REMOVE = "remove"


class Priority(str, Enum):
    """Priority level enum."""

    CRITICAL = "critical"
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


class UserSummary(BaseModel):
    """Minimal user info for embedded responses."""

    id: UUID
    name: str
    email: str


class TicketCreate(BaseModel):
    """Schema for creating a new ticket."""

    title: str = Field(..., min_length=5, max_length=200)
    description: str = Field(..., min_length=10)
    type: TicketType
    priority: Priority
    assigned_to: Optional[UUID] = None


class TicketUpdate(BaseModel):
    """Schema for updating a ticket (partial update)."""

    title: Optional[str] = Field(None, min_length=5, max_length=200)
    description: Optional[str] = Field(None, min_length=10)
    type: Optional[TicketType] = None
    priority: Optional[Priority] = None
    status: Optional[TicketStatus] = None
    assigned_to: Optional[UUID] = None


class TicketResponse(BaseModel):
    """Schema for ticket in API responses."""

    id: UUID
    project_id: UUID
    title: str
    description: str
    type: TicketType
    priority: Priority
    status: TicketStatus
    jira_key: Optional[str] = None
    created_by: Optional[UserSummary] = None
    assigned_to: Optional[UserSummary] = None
    attachments_count: int = 0
    activities_count: int = 0
    created_at: datetime
    updated_at: datetime


class TicketListResponse(BaseModel):
    """Schema for ticket list item (lighter than full response)."""

    id: UUID
    project_id: UUID
    title: str
    type: TicketType
    priority: Priority
    status: TicketStatus
    jira_key: Optional[str] = None
    created_by: Optional[UserSummary] = None
    assigned_to: Optional[UserSummary] = None
    created_at: datetime
    updated_at: datetime


class BulkStatusChange(BaseModel):
    """Schema for bulk status change operation."""

    ticket_ids: list[UUID] = Field(..., min_length=1, max_length=100)
    status: TicketStatus


class BulkAssign(BaseModel):
    """Schema for bulk assign operation."""

    ticket_ids: list[UUID] = Field(..., min_length=1, max_length=100)
    assigned_to: UUID


class BulkDelete(BaseModel):
    """Schema for bulk delete operation."""

    ticket_ids: list[UUID] = Field(..., min_length=1, max_length=100)